# Karol Kula projekt 1

## import bibliotek, załadowanie danych. Miasta na "K" nie było, więc wybrałem New Jersey

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

data = pd.read_csv("http://data.insideairbnb.com/united-states/nj/jersey-city/2023-09-18/data/listings.csv.gz")


# Wyświewtlenie 5 przykładówych wierszy, w celu zorientowania się co sie tu dzieje

In [2]:
data.head()

id                          listing_url       scrape_id last_scraped  \
0   40669   https://www.airbnb.com/rooms/40669  20230918050414   2023-09-18   
1  215768  https://www.airbnb.com/rooms/215768  20230918050414   2023-09-18   
2  254245  https://www.airbnb.com/rooms/254245  20230918050414   2023-09-18   
3  269266  https://www.airbnb.com/rooms/269266  20230918050414   2023-09-18   
4  270245  https://www.airbnb.com/rooms/270245  20230918050414   2023-09-18   

        source                                               name  \
0  city scrape  Condo in Jersey City · ★5.0 · 1 bedroom · 1 be...   
1  city scrape  Home in Jersey City · ★4.77 · 1 bedroom · 2 be...   
2  city scrape  Home in Jersey City · ★4.76 · 1 bedroom · 1 be...   
3  city scrape  Rental unit in Jersey City · ★4.36 · 2 bedroom...   
4  city scrape  Townhouse in Jersey City · ★4.55 · 1 bedroom ·...   

                                         description  \
0  <b>The space</b><br />Skyy’s Lounge ....Everyt...   
1  Walking to distance to Statue of Liberty and E...   
2  Walking to distance to Statue of Liberty and E...   
3  Enjoy that private luxury two bedrooms apartme...   
4  Just for you small bedroom with private bathro...   

                               neighborhood_overview  \
0  The neighborhood is very diverse & friendly sh...   
1  Such close proximity to NYC, 7 minutes on the ...   
2  Such close proximity to NYC, 7 minutes on the ...   
3  Our House is located in the Liberty State Park...   
4                                                NaN   

                                         picture_url  host_id  ...  \
0  https://a0.muscache.com/pictures/af7e4a45-0118...   175412  ...   
1  https://a0.muscache.com/pictures/miso/Hosting-...   846837  ...   
2  https://a0.muscache.com/pictures/miso/Hosting-...   846837  ...   
3  https://a0.muscache.com/pictures/miso/Hosting-...  1410590  ...   
4  https://a0.muscache.com/pictures/miso/Hosting-...  1410590  ...   

  review_scores_communication review_scores_location review_scores_value  \
0                        5.00                   5.00                5.00   
1                        4.87                   4.32                4.79   
2                        4.83                   4.32                4.74   
3                        4.66                   3.85                4.46   
4                        4.71                   3.99                4.59   

       license instant_bookable calculated_host_listings_count  \
0          NaN                f                              1   
1  STR20-00055                f                              2   
2  STR20-00055                f                              2   
3          NaN                t                             11   
4          NaN                t                             11   

  calculated_host_listings_count_entire_homes  \
0                                           0   
1                                           2   
2                                           2   
3                                           1   
4                                           1   

  calculated_host_listings_count_private_rooms  \
0                                            1   
1                                            0   
2                                            0   
3                                           10   
4                                           10   

  calculated_host_listings_count_shared_rooms reviews_per_month  
0                                           0              0.07  
1                                           0              1.17  
2                                           0              0.88  
3                                           0              2.91  
4                                           0              2.07  

[5 rows x 75 columns]

# Wyświetlenie wszystkich danych

In [3]:
pd.set_option('display.max_columns', None)
data

id                                      listing_url  \
0                  40669               https://www.airbnb.com/rooms/40669   
1                 215768              https://www.airbnb.com/rooms/215768   
2                 254245              https://www.airbnb.com/rooms/254245   
3                 269266              https://www.airbnb.com/rooms/269266   
4                 270245              https://www.airbnb.com/rooms/270245   
...                  ...                                              ...   
1319  980482250747395799  https://www.airbnb.com/rooms/980482250747395799   
1320  981050569727129557  https://www.airbnb.com/rooms/981050569727129557   
1321  981716811722485393  https://www.airbnb.com/rooms/981716811722485393   
1322  981829528456499146  https://www.airbnb.com/rooms/981829528456499146   
1323  982161671070648813  https://www.airbnb.com/rooms/982161671070648813   

           scrape_id last_scraped       source  \
0     20230918050414   2023-09-18  city scrape   
1     20230918050414   2023-09-18  city scrape   
2     20230918050414   2023-09-18  city scrape   
3     20230918050414   2023-09-18  city scrape   
4     20230918050414   2023-09-18  city scrape   
...              ...          ...          ...   
1319  20230918050414   2023-09-18  city scrape   
1320  20230918050414   2023-09-18  city scrape   
1321  20230918050414   2023-09-18  city scrape   
1322  20230918050414   2023-09-18  city scrape   
1323  20230918050414   2023-09-18  city scrape   

                                                   name  \
0     Condo in Jersey City · ★5.0 · 1 bedroom · 1 be...   
1     Home in Jersey City · ★4.77 · 1 bedroom · 2 be...   
2     Home in Jersey City · ★4.76 · 1 bedroom · 1 be...   
3     Rental unit in Jersey City · ★4.36 · 2 bedroom...   
4     Townhouse in Jersey City · ★4.55 · 1 bedroom ·...   
...                                                 ...   
1319  Rental unit in Jersey City · ★New · 1 bedroom ...   
1320  Rental unit in Jersey City · ★New · 1 bedroom ...   
1321  Rental unit in Jersey City · ★New · 1 bedroom ...   
1322  Home in Bayonne · ★New · 1 bedroom · 1 bed · 1...   
1323  Home in Jersey City · ★New · 3 bedrooms · 3 be...   

                                            description  \
0     <b>The space</b><br />Skyy’s Lounge ....Everyt...   
1     Walking to distance to Statue of Liberty and E...   
2     Walking to distance to Statue of Liberty and E...   
3     Enjoy that private luxury two bedrooms apartme...   
4     Just for you small bedroom with private bathro...   
...                                                 ...   
1319  Safe location, charming place , 2-3 min walkin...   
1320  Welcome to our inviting shared bedroom at Jers...   
1321         This unique place has a style all its own.   
1322  The room has an excellent location, it is a qu...   
1323  Lleva a toda la familia a este fantástico luga...   

                                  neighborhood_overview  \
0     The neighborhood is very diverse & friendly sh...   
1     Such close proximity to NYC, 7 minutes on the ...   
2     Such close proximity to NYC, 7 minutes on the ...   
3     Our House is located in the Liberty State Park...   
4                                                   NaN   
...                                                 ...   
1319                                                NaN   
1320  Our apartment is strategically located, making...   
1321                                                NaN   
1322                                                NaN   
1323                                                NaN   

                                            picture_url    host_id  \
0     https://a0.muscache.com/pictures/af7e4a45-0118...     175412   
1     https://a0.muscache.com/pictures/miso/Hosting-...     846837   
2     https://a0.muscache.com/pictures/miso/Hosting-...     846837   
3     https://a0.muscache.com/pictures/miso/Hosting-...    1410590   
4     https://a0.mu

# Wyswietlenie informacji o typach kolumn

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1324 entries, 0 to 1323
Data columns (total 75 columns):
 #   Column                                        Non-Null Count  Dtype  
---  ------                                        --------------  -----  
 0   id                                            1324 non-null   int64  
 1   listing_url                                   1324 non-null   object 
 2   scrape_id                                     1324 non-null   int64  
 3   last_scraped                                  1324 non-null   object 
 4   source                                        1324 non-null   object 
 5   name                                          1324 non-null   object 
 6   description                                   1317 non-null   object 
 7   neighborhood_overview                         819 non-null    object 
 8   picture_url                                   1324 non-null   object 
 9   host_id                                       1324 non-null   i

# Wyświetlenie rozmiaru zbioru danych

In [5]:
num_rows, num_cols = data.shape
print(num_rows, num_cols)

1324 75


# Mamy 75 kolumn i 1324 wiersze, wiekszość danych jest typu object więc będzie konieczna zamiana na konktretny odpowiedni typ. po wstępnym przeglądzie danych widać że niektóre kolumny są puste więc być może muszą zostać usunięte ponieważ nic nie wniosą do analizy, należy też sprawdzić czy nie można ich jakoś uzupełnić 

# Wyświetlenie samych typów kolumn

In [6]:
data_type = data.dtypes
data_type

id                                                int64
listing_url                                      object
scrape_id                                         int64
last_scraped                                     object
source                                           object
                                                 ...   
calculated_host_listings_count                    int64
calculated_host_listings_count_entire_homes       int64
calculated_host_listings_count_private_rooms      int64
calculated_host_listings_count_shared_rooms       int64
reviews_per_month                               float64
Length: 75, dtype: object

# Zamiana typów
## Teraz przeprowadze analize wszystkich kolumn i gdzie trzeba to zamienie na odpowiednie typy

In [7]:
data.iloc[:, :5]

id                                      listing_url  \
0                  40669               https://www.airbnb.com/rooms/40669   
1                 215768              https://www.airbnb.com/rooms/215768   
2                 254245              https://www.airbnb.com/rooms/254245   
3                 269266              https://www.airbnb.com/rooms/269266   
4                 270245              https://www.airbnb.com/rooms/270245   
...                  ...                                              ...   
1319  980482250747395799  https://www.airbnb.com/rooms/980482250747395799   
1320  981050569727129557  https://www.airbnb.com/rooms/981050569727129557   
1321  981716811722485393  https://www.airbnb.com/rooms/981716811722485393   
1322  981829528456499146  https://www.airbnb.com/rooms/981829528456499146   
1323  982161671070648813  https://www.airbnb.com/rooms/982161671070648813   

           scrape_id last_scraped       source  
0     20230918050414   2023-09-18  city scrape  
1     20230918050414   2023-09-18  city scrape  
2     20230918050414   2023-09-18  city scrape  
3     20230918050414   2023-09-18  city scrape  
4     20230918050414   2023-09-18  city scrape  
...              ...          ...          ...  
1319  20230918050414   2023-09-18  city scrape  
1320  20230918050414   2023-09-18  city scrape  
1321  20230918050414   2023-09-18  city scrape  
1322  20230918050414   2023-09-18  city scrape  
1323  20230918050414   2023-09-18  city scrape  

[1324 rows x 5 columns]

In [8]:
data['id'].duplicated().any()
data['source'].value_counts()

source
city scrape        1291
previous scrape      33
Name: count, dtype: int64

# typ int64 to dobry typ dla kolumny id, każde id jest unikalne
# listing_url to adres url ogłoszeń, można to zamienić na string
# scrape_id dobry typ
# last_scraped - moża zmienić na date
# source - można zmienić na kategorie


In [9]:
data['listing_url'] = data['listing_url'].astype('string')
data['listing_url'] = data['listing_url'].astype('string')
data['last_scraped'] = pd.to_datetime(data['last_scraped'])
data['source'] = data['source'].astype('category')

In [10]:
data.iloc[:, 5:10]

name  \
0     Condo in Jersey City · ★5.0 · 1 bedroom · 1 be...   
1     Home in Jersey City · ★4.77 · 1 bedroom · 2 be...   
2     Home in Jersey City · ★4.76 · 1 bedroom · 1 be...   
3     Rental unit in Jersey City · ★4.36 · 2 bedroom...   
4     Townhouse in Jersey City · ★4.55 · 1 bedroom ·...   
...                                                 ...   
1319  Rental unit in Jersey City · ★New · 1 bedroom ...   
1320  Rental unit in Jersey City · ★New · 1 bedroom ...   
1321  Rental unit in Jersey City · ★New · 1 bedroom ...   
1322  Home in Bayonne · ★New · 1 bedroom · 1 bed · 1...   
1323  Home in Jersey City · ★New · 3 bedrooms · 3 be...   

                                            description  \
0     <b>The space</b><br />Skyy’s Lounge ....Everyt...   
1     Walking to distance to Statue of Liberty and E...   
2     Walking to distance to Statue of Liberty and E...   
3     Enjoy that private luxury two bedrooms apartme...   
4     Just for you small bedroom with private bathro...   
...                                                 ...   
1319  Safe location, charming place , 2-3 min walkin...   
1320  Welcome to our inviting shared bedroom at Jers...   
1321         This unique place has a style all its own.   
1322  The room has an excellent location, it is a qu...   
1323  Lleva a toda la familia a este fantástico luga...   

                                  neighborhood_overview  \
0     The neighborhood is very diverse & friendly sh...   
1     Such close proximity to NYC, 7 minutes on the ...   
2     Such close proximity to NYC, 7 minutes on the ...   
3     Our House is located in the Liberty State Park...   
4                                                   NaN   
...                                                 ...   
1319                                                NaN   
1320  Our apartment is strategically located, making...   
1321                                                NaN   
1322                                                NaN   
1323                                                NaN   

                                            picture_url    host_id  
0     https://a0.muscache.com/pictures/af7e4a45-0118...     175412  
1     https://a0.muscache.com/pictures/miso/Hosting-...     846837  
2     https://a0.muscache.com/pictures/miso/Hosting-...     846837  
3     https://a0.muscache.com/pictures/miso/Hosting-...    1410590  
4     https://a0.muscache.com/pictures/miso/Hosting-...    1410590  
...                                                 ...        ...  
1319  https://a0.muscache.com/pictures/hosting/Hosti...  536701303  
1320  https://a0.muscache.com/pictures/miso/Hosting-...  513781380  
1321  https://a0.muscache.com/pictures/75b1cfdc-3b1d...   98240103  
1322  https://a0.muscache.com/pictures/hosting/Hosti...  333174348  
1323  https://a0.muscache.com/pictures/hosting/Hosti...  537476710  

[1324 rows x 5 columns]

# Wygląda na to że odpowiedni typ dla kolumn od name do picture_url to string 
# host_id ma odpowiedni typ

In [11]:
data['name'] = data['name'].astype('string')
data['description'] = data['description'].astype('string')
data['neighborhood_overview'] = data['neighborhood_overview'].astype('string')
data['picture_url'] = data['picture_url'].astype('string')

In [12]:
data.iloc[:, 10:15]

host_url  host_name  host_since  \
0        https://www.airbnb.com/users/show/175412       Skyy  2010-07-20   
1        https://www.airbnb.com/users/show/846837  Charlaine  2011-07-20   
2        https://www.airbnb.com/users/show/846837  Charlaine  2011-07-20   
3       https://www.airbnb.com/users/show/1410590      Magda  2011-11-15   
4       https://www.airbnb.com/users/show/1410590      Magda  2011-11-15   
...                                           ...        ...         ...   
1319  https://www.airbnb.com/users/show/536701303      Fares  2023-09-11   
1320  https://www.airbnb.com/users/show/513781380         Jc  2023-05-08   
1321   https://www.airbnb.com/users/show/98240103      Julio  2016-10-05   
1322  https://www.airbnb.com/users/show/333174348    Winston  2020-02-03   
1323  https://www.airbnb.com/users/show/537476710     Jose A  2023-09-16   

        host_location                                         host_about  
0     Jersey City, NJ  I am the owner of a high end Nail Salon in the...  
1     Jersey City, NJ  Thank you for taking the time to learn a bit a...  
2     Jersey City, NJ  Thank you for taking the time to learn a bit a...  
3     Jersey City, NJ  I am living in New York City area  for several...  
4     Jersey City, NJ  I am living in New York City area  for several...  
...               ...                                                ...  
1319              NaN                                                NaN  
1320  Jersey City, NJ  JC Livin is a small business that specializes ...  
1321  Jersey City, NJ  My Name is Julio Gallardo and I am married to ...  
1322              NaN                                                NaN  
1323  Jersey City, NJ                                                NaN  

[1324 rows x 5 columns]

In [13]:
data['host_location'].value_counts()

host_location
Jersey City, NJ              541
New York, NY                 273
Hawthorne, NJ                 39
Ridgewood, NJ                 35
New Jersey, United States     27
New York, United States       25
Weehawken, NJ                 10
Hoboken, NJ                   10
United States                  8
Los Angeles, CA                6
Edison, NJ                     5
Medellín, Colombia             5
West New York, NJ              4
Carlstadt, NJ                  4
Wayne, NJ                      4
Chicago, IL                    4
Ridgefield Park, NJ            2
Bayonne, NJ                    2
Harrison, NJ                   2
Keene Valley, NY               2
Las Vegas, NV                  2
Lackawaxen, PA                 2
Elizabeth, NJ                  2
Charlotte, NC                  1
Panama                         1
East Orange, NJ                1
Heidelberg, Germany            1
London, United Kingdom         1
Watertown, NY                  1
Holmdel, NJ                  

In [14]:
data['host_location'].value_counts().count()

56

# host_url, host_name, host_about - zamiana na typ string
# host_since - zamiana na date
# host_location - po sprawdzeniu ilości unikalnych wartości zdecydowałem się na typ string


In [15]:
data['host_url'] = data['host_url'].astype('string')
data['host_name'] = data['host_name'].astype('string')
data['host_since'] = pd.to_datetime(data['host_since'], errors='coerce')
data['host_location'] = data['host_location'].astype('string')
data['host_about'] = data['host_about'].astype('string')

In [16]:
data.iloc[:, 15:20]

host_response_time host_response_rate host_acceptance_rate  \
0          within a day                71%                  60%   
1        within an hour               100%                  99%   
2        within an hour               100%                  99%   
3        within an hour               100%                 100%   
4        within an hour               100%                 100%   
...                 ...                ...                  ...   
1319                NaN                NaN                 100%   
1320     within an hour               100%                  96%   
1321     within an hour               100%                  99%   
1322                NaN                NaN                  NaN   
1323                NaN                NaN                  NaN   

     host_is_superhost                                 host_thumbnail_url  
0                    f  https://a0.muscache.com/im/users/175412/profil...  
1                    t  https://a0.muscache.com/im/pictures/user/User-...  
2                    t  https://a0.muscache.com/im/pictures/user/User-...  
3                    f  https://a0.muscache.com/im/users/1410590/profi...  
4                    f  https://a0.muscache.com/im/users/1410590/profi...  
...                ...                                                ...  
1319                 f  https://a0.muscache.com/im/pictures/user/User-...  
1320                 f  https://a0.muscache.com/im/pictures/user/User-...  
1321                 f  https://a0.muscache.com/im/pictures/user/1918a...  
1322                 f  https://a0.muscache.com/im/pictures/user/ce737...  
1323                 f  https://a0.muscache.com/im/pictures/user/User-...  

[1324 rows x 5 columns]

In [17]:
data['host_response_time'].value_counts()

host_response_time
within an hour        1014
within a few hours     134
within a day            91
a few days or more      52
Name: count, dtype: int64

In [18]:
data['host_is_superhost'].value_counts()

host_is_superhost
f    920
t    384
Name: count, dtype: int64

In [19]:
data['host_acceptance_rate'].str.contains('%').all()

True

In [20]:
data['host_response_rate'].str.contains('%').all()

True

# host_response_time - tylko cztery wartości więc odpowiedni typ będzie kategoryczny
# host_response_rate, host_acceptance_rate wartości podane w procentach, podzielenie na sto i zmiana na typ float
# host_is_superhost - zawiera tylko wartości true i false  więc zamieniam na typ bool
# host_thumbnail_url - kolejny url a więc string

In [21]:
data['host_response_time'] = data['host_response_time'].astype('category')
data['host_response_rate'] = data['host_response_rate'].str.rstrip('%').astype('float') / 100
data['host_acceptance_rate'] = data['host_acceptance_rate'].str.rstrip('%').astype('float') / 100
data['host_is_superhost'] = data['host_is_superhost'].replace({'t': True, 'f': False}).astype(bool)
data['host_thumbnail_url'] = data['host_thumbnail_url'].astype('string')


In [22]:
data.iloc[:, 20:25]

host_picture_url    host_neighbourhood  \
0     https://a0.muscache.com/im/users/175412/profil...                   NaN   
1     https://a0.muscache.com/im/pictures/user/User-...                   NaN   
2     https://a0.muscache.com/im/pictures/user/User-...                   NaN   
3     https://a0.muscache.com/im/users/1410590/profi...      Bergen-Lafayette   
4     https://a0.muscache.com/im/users/1410590/profi...      Bergen-Lafayette   
...                                                 ...                   ...   
1319  https://a0.muscache.com/im/pictures/user/User-...  Downtown Jersey City   
1320  https://a0.muscache.com/im/pictures/user/User-...      Bergen-Lafayette   
1321  https://a0.muscache.com/im/pictures/user/1918a...           The Heights   
1322  https://a0.muscache.com/im/pictures/user/ce737...            Greenville   
1323  https://a0.muscache.com/im/pictures/user/User-...             West Side   

      host_listings_count  host_total_listings_count  host_verifications  
0                       1                          4  ['email', 'phone']  
1                       2                          4  ['email', 'phone']  
2                       2                          4  ['email', 'phone']  
3                      11                         18  ['email', 'phone']  
4                      11                         18  ['email', 'phone']  
...                   ...                        ...                 ...  
1319                    1                          2           ['phone']  
1320                    5                          5  ['email', 'phone']  
1321                   22                         32  ['email', 'phone']  
1322                    5                          6           ['phone']  
1323                    1                          1           ['phone']  

[1324 rows x 5 columns]

In [23]:
data['host_neighbourhood'].value_counts()

host_neighbourhood
The Heights                      243
Bergen-Lafayette                 159
Journal Square                   129
Downtown Jersey City             112
Greenville                        80
Cambridge                         79
West Side                         72
McGinley Square                   47
Hell's Kitchen                    38
The Village                       21
Hackensack River Waterfront       18
Upper East Side                   17
Hamilton Park                     14
Hudson Exchange                   11
East Colorado Springs             10
Newport                            8
Highland Lakes                     7
Bedford-Stuyvesant                 6
Powerhouse Arts District           6
Vegas de La Doctora                5
Iselin                             5
South Broad Street                 4
Potomac Yard                       3
Flushing                           3
North Riverdale                    3
Camelback East Village             2
Long Island City   

In [24]:
data['host_verifications'].value_counts()

host_verifications
['email', 'phone']                  912
['email', 'phone', 'work_email']    242
['phone']                           165
['phone', 'work_email']               4
['email']                             1
Name: count, dtype: int64

In [25]:
import ast
def parse_verification_list(verification_string):
    try:
        verification_list = ast.literal_eval(verification_string)
        if isinstance(verification_list, list):
            return verification_list
        else:
            return []

    except (ValueError, SyntaxError):
        return []

host_verifications_list = data['host_verifications'].apply(parse_verification_list)

In [26]:
host_verifications_list

0       [email, phone]
1       [email, phone]
2       [email, phone]
3       [email, phone]
4       [email, phone]
             ...      
1319           [phone]
1320    [email, phone]
1321    [email, phone]
1322           [phone]
1323           [phone]
Name: host_verifications, Length: 1324, dtype: object

# host_picture_url, host_picture_url mozna zamienić na typ string
# host_listings_count, host_total_listings_count mają odpowiednie typy
# host_verifications kolumna ta posiada liste, 3 wartości można to zamienić procesem One-hot encoding na osobne kolumny  tak aby każda wartości była oznaczona 1 lub 0 a kolumne host_verification można usunąć

In [27]:
data['host_picture_url'] = data['host_picture_url'].astype('string')
data['host_neighbourhood'] = data['host_neighbourhood'].astype('string')

from sklearn.preprocessing import MultiLabelBinarizer

mlb = MultiLabelBinarizer()

one_hot_encoded_verifications = mlb.fit_transform(host_verifications_list)
one_hot_encoded_df = pd.DataFrame(one_hot_encoded_verifications, columns=mlb.classes_)
one_hot_encoded_df.head()

data_with_one_hot = data.join(one_hot_encoded_df)

one_hot_encoded_verifications = mlb.fit_transform(host_verifications_list)
one_hot_encoded_df = pd.DataFrame(one_hot_encoded_verifications, columns=mlb.classes_)

data = pd.concat([data, one_hot_encoded_df], axis=1)
data.drop('host_verifications', axis=1, inplace=True)

In [28]:
data[['email', 'phone', 'work_email']]

email  phone  work_email
0         1      1           0
1         1      1           0
2         1      1           0
3         1      1           0
4         1      1           0
...     ...    ...         ...
1319      0      1           0
1320      1      1           0
1321      1      1           0
1322      0      1           0
1323      0      1           0

[1324 rows x 3 columns]

In [29]:
data.iloc[:, 24:30]

host_has_profile_pic host_identity_verified  \
0                       t                      f   
1                       t                      t   
2                       t                      t   
3                       t                      f   
4                       t                      f   
...                   ...                    ...   
1319                    t                      t   
1320                    t                      t   
1321                    t                      t   
1322                    t                      t   
1323                    t                      t   

                               neighbourhood  \
0     Jersey City, New Jersey, United States   
1     Jersey City, New Jersey, United States   
2     Jersey City, New Jersey, United States   
3     Jersey City, New Jersey, United States   
4                                        NaN   
...                                      ...   
1319                                     NaN   
1320  Jersey City, New Jersey, United States   
1321                                     NaN   
1322                                     NaN   
1323                                     NaN   

                           neighbourhood_cleansed  \
0         Ward C (councilmember Richard Boggiano)   
1     Ward F (councilmember Jermaine D. Robinson)   
2     Ward F (councilmember Jermaine D. Robinson)   
3     Ward F (councilmember Jermaine D. Robinson)   
4     Ward F (councilmember Jermaine D. Robinson)   
...                                           ...   
1319         Ward E (councilmember James Solomon)   
1320  Ward F (councilmember Jermaine D. Robinson)   
1321         Ward A (councilmember Denise Ridley)   
1322         Ward A (councilmember Denise Ridley)   
1323       Ward B (councilmember Mira Prinz-Arey)   

      neighbourhood_group_cleansed   latitude  
0                              NaN  40.737240  
1                              NaN  40.714380  
2                              NaN  40.711620  
3                              NaN  40.711350  
4                              NaN  40.713400  
...                            ...        ...  
1319                           NaN  40.719732  
1320                           NaN  40.705539  
1321                           NaN  40.688882  
1322                           NaN  40.687234  
1323                           NaN  40.732470  

[1324 rows x 6 columns]

In [30]:
data['host_has_profile_pic'].value_counts()

host_has_profile_pic
t    1291
f      33
Name: count, dtype: int64

In [31]:
data['host_identity_verified'].value_counts()

host_identity_verified
t    1200
f     124
Name: count, dtype: int64

In [32]:
data['neighbourhood'].value_counts()
#można wartością najcześciej występująca uzupełnić

neighbourhood
Jersey City, New Jersey, United States      815
North Bergen, New Jersey, United States       1
Hoboken, New Jersey, United States            1
Bayonne, New Jersey, United States            1
 jersey city , New Jersey, United States      1
Name: count, dtype: int64

In [33]:
data['neighbourhood_cleansed'].value_counts()

neighbourhood_cleansed
Ward E (councilmember James Solomon)           343
Ward C (councilmember Richard Boggiano)        252
Ward F (councilmember Jermaine D. Robinson)    246
Ward D (councilmember Michael Yun)             225
Ward B (councilmember Mira Prinz-Arey)         139
Ward A (councilmember Denise Ridley)           119
Name: count, dtype: int64

In [34]:
data['neighbourhood_group_cleansed'].value_counts()

Series([], Name: count, dtype: int64)

In [35]:
data['neighbourhood_group_cleansed'].isnull().all()

True

#  host_has_profile_pic, host_has_profile_pic typ bool
# neighbourhood, neighbourhood_cleansed ze względu na liczbe uniklanych wartości zamieniłem typ na category
# neighbourhood_group_cleansed - nie zmieniałem nic bo to jest pusta kolumna
# latitude - poprawny typ

In [36]:
data['host_has_profile_pic'] = data['host_has_profile_pic'].replace({'t': True, 'f': False}).astype(bool)
data['host_identity_verified'] = data['host_identity_verified'].replace({'t': True, 'f': False}).astype(bool)
data['neighbourhood'] = data['neighbourhood'].astype('category')
data['neighbourhood_cleansed'] = data['neighbourhood_cleansed'].astype('category')

In [37]:
data.iloc[:, 30:35]

longitude                property_type        room_type  accommodates  \
0    -74.053100        Private room in condo     Private room             2   
1    -74.058100                  Entire home  Entire home/apt             4   
2    -74.061790                  Entire home  Entire home/apt             2   
3    -74.063740           Entire rental unit  Entire home/apt             5   
4    -74.064890    Private room in townhouse     Private room             1   
...         ...                          ...              ...           ...   
1319 -74.046680  Private room in rental unit     Private room             1   
1320 -74.076724  Private room in rental unit     Private room             2   
1321 -74.097413  Private room in rental unit     Private room             2   
1322 -74.099546         Private room in home     Private room             2   
1323 -74.075320                  Entire home  Entire home/apt             8   

      bathrooms  
0           NaN  
1           NaN  
2           NaN  
3           NaN  
4           NaN  
...         ...  
1319        NaN  
1320        NaN  
1321        NaN  
1322        NaN  
1323        NaN  

[1324 rows x 5 columns]

In [38]:
data['property_type'].value_counts()

property_type
Entire rental unit                    489
Private room in home                  187
Private room in rental unit           184
Entire serviced apartment             115
Entire home                            93
Entire condo                           93
Private room in condo                  25
Private room in serviced apartment     23
Entire townhouse                       22
Private room in townhouse              20
Private room                           18
Entire loft                            11
Room in hotel                          11
Entire guest suite                     10
Entire guesthouse                       6
Private room in loft                    5
Private room in guest suite             2
Entire vacation home                    2
Private room in bed and breakfast       1
Entire home/apt                         1
Tiny home                               1
Boat                                    1
Room in boutique hotel                  1
Shared room in bed a

In [39]:
data['room_type'].value_counts()

room_type
Entire home/apt    845
Private room       476
Shared room          2
Hotel room           1
Name: count, dtype: int64

#  longitude zostawiam jako float 
# property_type zamieniam na category
# room_type zamieniam na category
# accommodates zostawiam int64
# bathrooms zostawiam jak jest bo jest puste

In [40]:
data['property_type'] = data['property_type'].astype('category')
data['room_type'] = data['room_type'].astype('category')

In [41]:
data.dtypes

id                                                      int64
listing_url                                    string[python]
scrape_id                                               int64
last_scraped                                   datetime64[ns]
source                                               category
                                                    ...      
calculated_host_listings_count_shared_rooms             int64
reviews_per_month                                     float64
email                                                   int32
phone                                                   int32
work_email                                              int32
Length: 77, dtype: object

In [42]:
data['bathrooms_text'] = data['bathrooms_text'].astype('string')
data['amenities'] = data['amenities'].astype('string')

data['price'] = data['price'].str.replace(r'[^0-9.]', '', regex=True).astype(float)



# Check if 'bedrooms' and 'beds' contain only integers (or NaN)

unique_bedrooms = data['bedrooms'].dropna().unique()

unique_beds = data['beds'].dropna().unique()

is_bedrooms_integers = all(x.is_integer() for x in unique_bedrooms)

is_beds_integers = all(x.is_integer() for x in unique_beds)



# Convert 'bedrooms' and 'beds' to int types if they contain only integer values

if is_bedrooms_integers:

    data['bedrooms'] = data['bedrooms'].astype('Int64')  # Use 'Int64' (capitalized) to allow NaN values

if is_beds_integers:

    data['beds'] = data['beds'].astype('Int64')  # Use 'Int64' (capitalized) to allow NaN values




In [43]:
# Convert 'has_availability' to bool type after checking values

data['has_availability'] = data['has_availability'].replace({'t': True, 'f': False}).astype(bool)



# Convert 'calendar_last_scraped' to datetime type

data['calendar_last_scraped'] = pd.to_datetime(data['calendar_last_scraped'], errors='coerce')



# Convert 'first_review' and 'last_review' to datetime type

data['first_review'] = pd.to_datetime(data['first_review'], errors='coerce')

data['last_review'] = pd.to_datetime(data['last_review'], errors='coerce')



# Check if 'calendar_updated' is a column with dates or needs to be dropped

calendar_updated_unique_values = data['calendar_updated'].unique()

# If all values are NaN, we assume it's not useful and could be dropped

if len(calendar_updated_unique_values) == 1 and pd.isna(calendar_updated_unique_values[0]):

    data.drop('calendar_updated', axis=1, inplace=True)
